File: src/internal/domain/repository_interface/user_repository.rs
use async_trait::async_trait;
use uuid::Uuid;
use crate::{
    dto::{
        res::{
            token::token_user_response::TokenUserResponse,
            user_response::UserResponse,
        },
    },
    internal::domain::entities::user_entity::{ChangeUserEntity, UserEntity},
};

#[async_trait]
pub trait UserRepository {
    fn new() -> Self where Self: Sized;
    async fn find_by_email(&self, email: String) -> Result<bool, String>;
    async fn find_by_id(&self, id: Uuid) -> Result<bool, String>;
    async fn check_user_role(&self, id: Uuid) -> Result<String, String>;
    async fn sign_up(&self, user: UserEntity) -> Result<(), String>;
    async fn get_hashed_password_with_salt(&self, email: String) -> Result<(Vec<u8>, String), String>;
    async fn log_in(&self, email: String, password: Vec<u8>) -> Result<TokenUserResponse, String>;
    async fn get_user(&self, id: Uuid) -> Result<UserResponse, String>;
    async fn change_user(&self, user: ChangeUserEntity) -> Result<(), String>;
}

File: internal/domain/repo_interface/user_repository.go
package repo_interface

import (
	"blueberry_homework/dto/response"
	"blueberry_homework/internal/domain/entities"

	"github.com/gocql/gocql"
)

// UserRepository는 사용자를 관리하는 인터페이스입니다.
type UserRepository interface {
	// 이메일로 사용자 찾기 함수
	FindByEmail(email string) (bool, error)

	// id로 사용자 찾기 함수
	FindById(id gocql.UUID) (bool, error)

	// 유저의 역할을 확인하는 함수
	CheckRole(id gocql.UUID) (string, error)

	// 회원가입 함수
	SignUp(entity entities.UserEntity) error

	// 해싱된 비밀번호 가져오는 함수
	GetHashedPassword(email string) (string, error)

	// 토큰 정보 가져오는 함수
	GetTokenInfo(email string) (string, string, error)

	// GetUser는 유저의 정보를 가져옵니다
	GetUser(id gocql.UUID) (response.UserResponse, error)

	// 사용자 정보 변경 함수
	ChangeUser(user entities.ChangeUserEntity) error
}
